client/src/main/java/org/eclipse/hono/client/impl/TenantClientImpl.java

```diff
@@ -0,0 +1,155 @@
+/**
+ * Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 1.0 which is available at
+ * https://www.eclipse.org/legal/epl-v10.html
+ *
+ * SPDX-License-Identifier: EPL-1.0
+ */
+
+package org.eclipse.hono.client.impl;
+
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.UUID;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.proton.ProtonReceiver;
+import io.vertx.proton.ProtonSender;
+import org.eclipse.hono.client.StatusCodeMapper;
+import org.eclipse.hono.client.TenantClient;
+import org.eclipse.hono.config.ClientConfigProperties;
+
+import org.eclipse.hono.util.TenantConstants;
+import org.eclipse.hono.util.TenantObject;
+import org.eclipse.hono.util.TenantResult;
+import org.eclipse.hono.util.MessageHelper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import io.vertx.core.Context;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.proton.ProtonConnection;
+
+/**
+ * A Vertx-Proton based client for Hono's Tenant API.
+ *
+ */
+public final class TenantClientImpl extends AbstractRequestResponseClient<TenantResult<TenantObject>>
```

Comment:
Can you not make it final in order to extend it. Instead, you can make existing methods final and allow concrete implementations to provide further functionalities. See PR #512 for similar extensions.

Reply:
Good point, was written in parallel :-)
Will remove the class final :+1: